### ISSUE TITLE
Can't distinguish "no output yet" from "no output ever" in layout mode

### ISSUE THREAD

**Issue by parente** [jupyter-incubator/dashboards#131]
From working with @vinomaster on a dashboard notebook:

1. Develop a notebook.
2. Run it top to bottom.
3. Go to layout mode the first time.
4. Try to identify cells that *will* have output after some condition is met (e.g., some series of widgets are clicked) to click "+" on them and lay them out.
5. Struggle because cells that will have output are indistinguishable from cells that will never have output down in the hidden area: they're both blank white boxes.

Solutions:

1. Tell user to interact with the notebook until all output shows up.
2. In hidden mode, show cell input if cell has no output (so that the code acts as an identifying feature).
3. Other?

**Comment by dalogsdon** [comment]
I would say it is a requirement that the cells be run before dashboard layout is used. We could detect if anything has been run (using metadata?) and put a notice in the layout view if nothing has been run or if something has not been run.

Showing the code could be good for identification, or maybe at least a snippet. Perhaps if an individual cell has not been run, the dashboard cell can show a "Run" button along with a snippet of code?

**Comment by parente** [comment]
Running all the cells is not enough. In @vinomaster's case, he has to run all cells **and** interact with every widget to ensure all output is rendered.

> Perhaps if an individual cell has not been run, the dashboard cell can show a "Run" button along with a snippet of code?

How about simpler: if a cell is in hidden and has no output, show its input. If the cell has output, show its output.

**Comment by dalogsdon** [comment]
> How about simpler: if a cell is in hidden and has no output, show its input. If the cell has output, show its output.

Should there be a distinction if the cell has not been run? It is possible to run from layout view, but there is no indication which cell is selected.

**Comment by jhpedemonte** [comment]
One option: when hovering over cells, show the code, making it easier for the user to distinguish cells with no (current) output:

![screen shot 2015-11-18 at 11 03 12 am](https://cloud.githubusercontent.com/assets/711365/11248165/4aa10668-8de5-11e5-9a5e-44cc3a60d78d.png)

Biggest downside is the code would still appear under the output in some cases:

![screen shot 2015-11-18 at 11 05 15 am](https://cloud.githubusercontent.com/assets/711365/11248181/6133dc52-8de5-11e5-984b-1489a292ac39.png)

But when moving mouse away from cells/dashboard, the code would disappear.  Thoughts?

**Comment by jhpedemonte** [comment]
To go along with my previous proposal: we currently have a toggle for the **hidden** cells to show/hide the code. We could move that up to the main dashboard as well and default to **on**, but allow the user to always hide the code if they choose.

**Comment by jhpedemonte** [comment]
> if a cell is in hidden and has no output, show its input. If the cell has output, show its output.

This is not easy to detect. For example, in the case of the taxi demo, after having run all of the cells, some cells display no output **visually**, but there is in fact DOM in the output area (waiting to be filled in, usually). I can't think of a good 100% accurate way of detecting this.

In fact, Drew and I took a stab at this a while back: when a user first goes to the Dashboard Layout view (to create a dashboard for the first time in that notebook), the code tries to only add the cells with output into the dashboard view at top, adding the remaining cells to the **hidden cells** area at the bottom. But because of what I explained in the previous paragraph, we couldn't detect which cells had output (or would have output) with 100% certainty.

In the end, we played it conservatively -- most "empty" cells end up in the **hidden** area, but some will also appear in dashboard, requiring the user to hide them if necessary.

**Comment by jtyberg** [comment]
> How about simpler: if a cell is in hidden and has no output, show its input. If the cell has output, show its output.

I like this suggestion, but if it's too difficult to distinguish whether or not a cell has output, then can we make it even simpler?  If a cell is hidden, show it's input and output, if any. 

Currently, seeing a ton of empty cells while scrolling through the hidden cells leaves me a bit confused.  IMHO, showing as much as possible of hidden cells gives me way more information that I can use while I'm quickly trying to determine what should be shown.

Put simply, more information about hidden cells is better.  I'm not sure I want to hover on each cell to figure out what's in it.  When I'm trying to determine what hidden cells to show, I'm usually scanning the cells visually.  I do not want this to be a thought intensive exercise.

**Comment by parente** [comment]
>  If a cell is hidden, show it's input and output, if any.

And if it's in the dashboard layout but has no output, just keep it blank? I'm OK with that. Checking if that's OK with you.

**Comment by jtyberg** [comment]
I guess it comes down to: should we ever show a blank cell in Layout mode.  Does a blank cell give me any information?  I think there should be SOMETHING in a cell.  I would rather see input AND output on top of each other in a cell than nothing at all.

**Comment by vinomaster** [comment]
So I think @jtyberg understands the pain-points here. We are providing a Layout view whereby the user needs to distinguish what cells are shown or hidden and where they go in the dashboard layout. This is impossible without some type of identifier. 

Ideally the ability to name cells would be nice. This is exactly what Excel users do when they need to reference a cell. Cleary this is a Jupyter core change but that to me is the proper fix. I would only name those of interest and could then find them in layout mode. This would be a step above what we did in KAWB with dashboard checkbox.

If it is NOT POSSIBLE to NAME CELLS then my thoughts are as follows:

1. All cells are by default HIDDEN.
2. Cells without output at design time should show the same content as shown when HIDDEN.
3. Hidden cells without output should show their input. It could be the capped at like no more than 10 lines assuming this is adequate for identification.

As for the notion of showing input content via hover-over, that seems like a nice thing but i do not see that as a requirement.

**Comment by jhpedemonte** [comment]
There seems to be some slight confusion with my proposal above. I've attached some animated GIFs to better help explain:

### Show code in dashboard area

![cell-code-hover](https://cloud.githubusercontent.com/assets/711365/11307009/05d6fb4a-8f7b-11e5-8fe3-71973dccd481.gif)

In the **main dashboard area**, when a user hovers over the whole area (not just individual cells), then the code for each cell is shown (in addition to the cell boundaries). Hovering off of the dashboard area allows the user to get immediate feedback as to what their layout will look like, without cell boundaries and without code text.

If a user finds this feature annoying, they can disable it with the checkbox in the top-right corner.

### Show code in hidden cells area

![hidde-cell-show-code](https://cloud.githubusercontent.com/assets/711365/11307058/564c07a0-8f7b-11e5-8750-dcb16c745b94.gif)

We have always had a **Show Code** button in the **Hidden Cells** area. When clicked, it will show the full input for each cell (in addition to output).

**Comment by jtyberg** [comment]
@jhpedemonte animated GIFs FTW!  I like the behavior.  

In the spirit of my previous comment (a blank cell doesn't tell you anything), is it worth considering showing the code in hidden cells by default?  Or is there a downside to doing so?

**Comment by jhpedemonte** [comment]
Now that I think about it, it doesn't really help to hide the code in the hidden area. So maybe we can change to just always show the code in the hidden area and remove the **Show/Hide code** button.

The only change I would consider making is to cap the height of the code/input areas, so the hidden area isn't terribly long. We just want to show the user enough to know what each code is.

@parente @dalogsdon: what do you think of this proposal?

**Comment by parente** [comment]
:yes: to always show the code in hidden. :yes: to capping the input height so that the hidden area doesn't become unwieldy and hard to drag from in layout mode.

**Comment by dalogsdon** [comment]
Always showing the code in the hidden area is OK. Since the hidden cells are pretty small vertically, should we:
1. make the code be scrollable
2. make hidden cells have a default height but also a little bit larger max height
3. leave it
?

**Comment by parente** [comment]
I'm OK with any implementation of "capping the height" I think.

**Comment by parente** [comment]
I merged the current PR after testing to avoid having new conflicts crop up. The hidden cell area changes can be put in via a second PR.
